### PR TITLE
Add SQL rendering tests for pushdown edge cases

### PR DESCRIPTION
**Pull request by tlento**
Add SQL rendering tests for pushdown edge cases

The predicate pushdown optimizer tests cover direct edge cases
for the optimizer alone, but we would like to have query rendering
variations of this so we can see how optimizers interact in their
query output.

This adds the relevant query rendering test scenarios along with
their associated snapshots.

Update SQL engine snapshots

### CODE REVIEW

**Comment by tlento** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1305?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1308** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1308?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1307** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1307?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1306** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1306?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1305** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1305?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1304** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1304?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1303** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1303?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1302** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1302?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1301** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1301?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1300** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1300?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1286** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1286?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1280** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1280?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1279** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1279?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1278** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1278?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1277** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1277?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1263** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1263?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1262** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1262?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1240** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1240?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1227** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1227?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1230** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1230?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1226** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1226?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1225** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1225?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1224** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1224?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1218** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1218?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1216** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1216?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1217](https://github.com/dbt-labs/metricflow/pull/1217) <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1217?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1215** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1215?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1214** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1214?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @tlento and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: tests_metricflow/snapshots/test_predicate_pushdown_rendering.py/SqlQueryPlan/DuckDB/test_cumulative_metric_with_query_time_filters__plan0_optimized.sql [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  -- Join Self Over Time Range
+  -- Pass Only Elements: ['bookers', 'booking__is_instant', 'metric_time__day', 'listing']
+  SELECT
+    subq_18.ds AS metric_time__day
+    , subq_16.listing AS listing
+    , subq_16.bookers AS bookers
+  FROM ***************************.mf_time_spine subq_18
+  INNER JOIN (
+    -- Constrain Output with WHERE
+    SELECT
```
Comment (courtneyholcomb) [inline tests_metricflow/snapshots/test_predicate_pushdown_rendering.py/SqlQueryPlan/DuckDB/test_cumulative_metric_with_query_time_filters__plan0_optimized.sql:20]: Would love if this subquery was combined with the inner subquery... but alas, a problem for a future date. Do we have a task up to fix that ?

File: tests_metricflow/snapshots/test_predicate_pushdown_rendering.py/SqlQueryPlan/DuckDB/test_fill_nulls_time_spine_metric_predicate_pushdown__plan0_optimized.sql [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            ) subq_50
+            ON
+              subq_52.ds - INTERVAL 14 day = subq_50.metric_time__day
+          ) subq_54
+          LEFT OUTER JOIN
+            ***************************.dim_listings_latest listings_latest_src_28000
+          ON
+            subq_54.listing = listings_latest_src_28000.listing_id
+        ) subq_59
+        WHERE booking__is_instant
```
Comment (courtneyholcomb) [inline tests_metricflow/snapshots/test_predicate_pushdown_rendering.py/SqlQueryPlan/DuckDB/test_fill_nulls_time_spine_metric_predicate_pushdown__plan0_optimized.sql:122]: Is it expected that this doesn't get pushed down further? (i.e., into the subquery that directly selects from `bookings_source_src_28000`)

File: tests_metricflow/snapshots/test_predicate_pushdown_rendering.py/SqlQueryPlan/DuckDB/test_metric_time_filter_with_two_targets__plan0_optimized.sql [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+      , listing_id AS listing
+      , 1 AS bookings
+    FROM ***************************.fct_bookings bookings_source_src_28000
+  ) subq_13
+  LEFT OUTER JOIN
+    ***************************.dim_listings_latest listings_latest_src_28000
+  ON
+    subq_13.listing = listings_latest_src_28000.listing_id
+) subq_18
+WHERE metric_time__day = '2024-01-01'
```
Comment (courtneyholcomb) [inline tests_metricflow/snapshots/test_predicate_pushdown_rendering.py/SqlQueryPlan/DuckDB/test_metric_time_filter_with_two_targets__plan0_optimized.sql:30]: This doesn't look like it got pushed down - is that expected?

File: tests_metricflow/snapshots/test_predicate_pushdown_rendering.py/SqlQueryPlan/DuckDB/test_simple_join_to_time_spine_pushdown_filter_application__plan0_optimized.sql [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    ) subq_10
+    WHERE booking__is_instant
+    GROUP BY
+      metric_time__day
+      , booking__is_instant
+  ) subq_13
+  ON
+    subq_15.ds = subq_13.metric_time__day
+) subq_16
+WHERE booking__is_instant
```
Comment (courtneyholcomb) [inline tests_metricflow/snapshots/test_predicate_pushdown_rendering.py/SqlQueryPlan/DuckDB/test_simple_join_to_time_spine_pushdown_filter_application__plan0_optimized.sql:39]: Reminding myself why we need this second where filter. I think it's because the time spine join might add null rows where `booking__is_instant = null`.

File: tests_metricflow/snapshots/test_predicate_pushdown_rendering.py/SqlQueryPlan/DuckDB/test_fill_nulls_time_spine_metric_predicate_pushdown__plan0_optimized.sql (hunk 2 above)
Comment (tlento) [reply tests_metricflow/snapshots/test_predicate_pushdown_rendering.py/SqlQueryPlan/DuckDB/test_fill_nulls_time_spine_metric_predicate_pushdown__plan0_optimized.sql:122]: Ah yes, because there's an offset in here. This one looks like it can be pushed down. I'm pretty sure I can relax that restriction now.

File: tests_metricflow/snapshots/test_predicate_pushdown_rendering.py/SqlQueryPlan/DuckDB/test_metric_time_filter_with_two_targets__plan0_optimized.sql (hunk 3 above)
Comment (tlento) [reply tests_metricflow/snapshots/test_predicate_pushdown_rendering.py/SqlQueryPlan/DuckDB/test_metric_time_filter_with_two_targets__plan0_optimized.sql:30]: Yes, time dimensions are not pushed down yet.

**Review by courtneyholcomb — APPROVED** [review]
Sounds like all the pushdown scenarios are expected (as discussed async) - looks good!

File: tests_metricflow/snapshots/test_predicate_pushdown_rendering.py/SqlQueryPlan/DuckDB/test_simple_join_to_time_spine_pushdown_filter_application__plan0_optimized.sql (hunk 4 above)
Comment (tlento) [reply tests_metricflow/snapshots/test_predicate_pushdown_rendering.py/SqlQueryPlan/DuckDB/test_simple_join_to_time_spine_pushdown_filter_application__plan0_optimized.sql:39]: Yes, that's correct.

**Comment by tlento** [comment]
### Merge activity

* **Jun 25, 5:24 PM PDT**: @tlento started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dbt-labs/metricflow/1305).
